Remove `token::Lit` from `ast::MetaItemLit`.

nnethercote opened the pull request:
Currently `ast::MetaItemLit` represents the literal kind twice. This PR removes that redundancy. Best reviewed one commit at a time.

r? @petrochenkov

nnethercote committed:
Rename `LitKind::to_token_lit` as `LitKind::synthesize_token_lit`.

This makes it clearer that it's not a lossless conversion, which I find
helpful.

nnethercote committed:
Add `StrStyle` to `ast::LitKind::ByteStr`.

This is required to distinguish between cooked and raw byte string
literals in an `ast::LitKind`, without referring to an adjacent
`token::Lit`. It's a prerequisite for the next commit.

nnethercote committed:
Remove `token::Lit` from `ast::MetaItemLit`.

`token::Lit` contains a `kind` field that indicates what kind of literal
it is. `ast::MetaItemLit` currently wraps a `token::Lit` but also has
its own `kind` field. This means that `ast::MetaItemLit` encodes the
literal kind in two different ways.

This commit changes `ast::MetaItemLit` so it no longer wraps
`token::Lit`. It now contains the `symbol` and `suffix` fields from
`token::Lit`, but not the `kind` field, eliminating the redundancy.

nnethercote committed:
Reorder `StrLit` fields.

To better match `MetaItemLit`.

petrochenkov commented on compiler/rustc_ast/src/attr/mod.rs:
```diff
@@ -328,7 +328,9 @@ pub fn mk_name_value_item_str(ident: Ident, str: Symbol, str_span: Span) -> Meta
 
 pub fn mk_name_value_item(ident: Ident, kind: LitKind, lit_span: Span) -> MetaItem {
-    let lit = MetaItemLit { token_lit: kind.to_token_lit(), kind, span: lit_span };
+    let token_lit = kind.synthesize_token_lit();
```
This use of `synthesize_token_lit` is suspicious, we don't need to go through a token if we are directly constructing a `MetaItemLit`, need to check where is this `mk_name_value_item` used.

nnethercote replied:
Removed.

petrochenkov commented on compiler/rustc_ast/src/attr/mod.rs:
```diff
@@ -408,7 +410,7 @@ pub fn mk_attr_name_value_str(
     span: Span,
 ) -> Attribute {
-    let lit = LitKind::Str(val, StrStyle::Cooked).to_token_lit();
+    let lit = LitKind::Str(val, StrStyle::Cooked).synthesize_token_lit();
```
Also suspicious for the same reason as `mk_name_value_item`.

nnethercote replied:
Removed.

petrochenkov commented on compiler/rustc_expand/src/build.rs:
```diff
@@ -333,7 +333,7 @@ impl<'a> ExtCtxt<'a> {
 
     fn expr_lit(&self, span: Span, lit_kind: ast::LitKind) -> P<ast::Expr> {
-        let token_lit = lit_kind.to_token_lit();
+        let token_lit = lit_kind.synthesize_token_lit();
```
Also suspicious, if an AST builder  function produces a token lit, it shouldn't take lowered `ast::LitKind`.

nnethercote replied:
Removed.

petrochenkov commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -1551,7 +1551,7 @@ impl<'a> Parser<'a> {
             });
             consume_colon = false;
-            Ok(self.mk_expr(lo, ExprKind::Lit(lit.token_lit)))
+            Ok(self.mk_expr(lo, ExprKind::Lit(lit.as_token_lit())))
```
Also suspicious, if we are in the parser we should not use `MetaItemLit`.

nnethercote replied:
Fixed.

petrochenkov commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -1817,7 +1818,7 @@ impl<'a> Parser<'a> {
         self.parse_opt_token_lit()
             .ok_or(())
-            .or_else(|()| self.handle_missing_lit().map(|lit| (lit.token_lit, lit.span)))
+            .or_else(|()| self.handle_missing_lit().map(|lit| (lit.as_token_lit(), lit.span)))
```
Same as in the previous comment.

nnethercote replied:
Fixed.

petrochenkov commented on compiler/rustc_parse/src/parser/pat.rs:
```diff
@@ -420,7 +420,7 @@ impl<'a> Parser<'a> {
                 err
             });
-            PatKind::Lit(self.mk_expr(lo, ExprKind::Lit(lit.token_lit)))
+            PatKind::Lit(self.mk_expr(lo, ExprKind::Lit(lit.as_token_lit())))
```
Same as in the previous comment.

nnethercote replied:
Fixed.

petrochenkov commented on compiler/rustc_ast/src/util/literal.rs:
```diff
@@ -196,9 +202,32 @@ impl LitKind {
+    }
+
+    /// Cheaply converts a meta item literal into a token literal.
+    pub fn as_token_lit(&self) -> token::Lit {
```
Basically, the only non-suspicious uses of `synthesize_token_lit` and this function are AST/HIR pretty-printing (including rustfmt), and recovering a token from `ExprKind::IncludedBytes`.

nnethercote replied:
I ended up removing `synthesize_token_lit`.

nnethercote committed:
Remove `mk_name_value_item{,_str}`.

There are better ways to create the meta items.
- In the rustdoc tests, the commit adds `dummy_meta_item_name_value`,
  which matches the existing `dummy_meta_item_word` function and
  `dummy_meta_item_list` macro.
- In `types.rs` the commit clones the existing meta item and then
  modifies the clone.

nnethercote committed:
Remove `ExtCtxt::expr_lit`.

nnethercote committed:
Parameterise `Parser::{recover_unclosed_char,handle_missing_lit}`.

These two methods both produce a `MetaItemLit`, and then some of the
call sites convert the `MetaItemLit` to a `token::Lit` with
`as_token_lit`.

This commit parameterises these two methods with a `mk_lit_char`
closure, which can be used to produce either `MetaItemLit` or
`token::Lit` directly as necessary.

nnethercote committed:
Remove three uses of `LitKind::synthesize_token_lit`.

nnethercote committed:
Remove `LitKind::synthesize_token_lit`.

It has a single call site in the HIR pretty printer, where the resulting
token lit is immediately converted to a string.

This commit replaces `LitKind::synthesize_token_lit` with a `Display`
impl for `LitKind`, which can be used by the HIR pretty printer.

nnethercote commented:
I have added five new commits addressing the review comments.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f -->
<!-- homu: {"type":"Approved","sha":"7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f with merge f9cf4381d57f19593c5669da52315f7dce06b33f...
<!-- homu: {"type":"BuildStarted","head_sha":"7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f","merge_sha":"f9cf4381d57f19593c5669da52315f7dce06b33f"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3637431436/jobs/6138750140)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3637431436/jobs/6138750140","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3637431436/jobs/6138441055) [(plain)](https://github.com/rust-lang-ci/rust/commit/f9cf4381d57f19593c5669da52315f7dce06b33f/checks/9936926071/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: 5870 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
thread 'main' panicked at 'assertion failed: status.success()', src/tools/cargotest/main.rs:148:13
Build completed unsuccessfully in 0:15:54
Build completed unsuccessfully in 0:15:54
make: *** [Makefile:44: check-aux] Error 1
```

</details>

bors commented:
:hourglass: Testing commit 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f with merge 8f76bdcad23479d2f4dd5fdf3520190e08ecded7...
<!-- homu: {"type":"BuildStarted","head_sha":"7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f","merge_sha":"8f76bdcad23479d2f4dd5fdf3520190e08ecded7"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3644541458/jobs/6154500655)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3644541458/jobs/6154500655","builder_name":"checks-actions"} -->

petrochenkov commented:
The `src/test\run-make\coverage-reports` failure may be legitimate.
@rustbot author

nnethercote commented:
@richkadel: this PR makes a few changes to the AST. Nothing especially notable, and the kind of thing I've done many times before. When merging there was a single coverage test failure, but only on Windows:
```
running 76 tests
test [run-make] src/test\run-make\coverage-reports ... FAILED
<...many successful and ignored test result lines omitted...>

failures:

---- [run-make] src/test\run-make\coverage-reports stdout ----

error: make failed
status: exit code: 2
command: "make"
--- stdout -------------------------------
make[1]: Entering directory '/d/a/rust/rust/src/test/run-make/coverage-reports'
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make/coverage-reports'
------------------------------------------
--- stderr -------------------------------
Makefile:83: clear_expected_if_blessed: No such file or directory
------------------------------------------

failures:
    [run-make] src/test\run-make\coverage-reports
```
Any idea what the problem might be? I have previously seen coverage failures with quite descriptive failure messages, but this one just has the `clear_expected_if_blessed` line. I admit I don't even understand what that part of the Makefile is doing.

nnethercote commented:
[This Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/run-make.2Fcoverage-reports) indicates that this `coverage-reports` failure is happening to other people as well.

richkadel commented:
Good to know it's unrelated to your change. `make` is really not good but it was the only tool available other than writing something from scratch. I haven't been working in the Rust codebase in a while so hopefully someone with more recent/relevant knowledge will figure out and fix the Makefile compatibility issues.

nnethercote commented:
#105477 has hopefully fixed the `coverage-reports` failure.

@bors r=petrochenkov

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #105456

bors commented:
:pushpin: Commit 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f -->
<!-- homu: {"type":"Approved","sha":"7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3644541458/jobs/6153894223) [(plain)](https://github.com/rust-lang-ci/rust/commit/8f76bdcad23479d2f4dd5fdf3520190e08ecded7/checks/9957416859/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

test result: FAILED. 46 passed; 1 failed; 29 ignored; 0 measured; 0 filtered out; finished in 8.07s

Build completed unsuccessfully in 0:55:18
make: *** [Makefile:73: ci-subset-1] Error 1
```

</details>

bors commented:
:hourglass: Testing commit 7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f with merge 2cd2070af7643ad88d280a4933bc4fb60451e521...
<!-- homu: {"type":"BuildStarted","head_sha":"7e0c6dba0d83dbee96bbf7eac7b4cb563e297a5f","merge_sha":"2cd2070af7643ad88d280a4933bc4fb60451e521"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3672911314/jobs/6211171560)
Approved by: petrochenkov
Pushing 2cd2070af7643ad88d280a4933bc4fb60451e521 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3672911314/jobs/6211171560"},"merge_sha":"2cd2070af7643ad88d280a4933bc4fb60451e521"} -->

rust-timer commented:
Finished benchmarking commit (2cd2070af7643ad88d280a4933bc4fb60451e521): [comparison URL](https://perf.rust-lang.org/compare.html?start=b397bc0727ad27340466166455c6edd327a589c4&end=2cd2070af7643ad88d280a4933bc4fb60451e521&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b397bc0727ad27340466166455c6edd327a589c4&end=2cd2070af7643ad88d280a4933bc4fb60451e521&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [1.0%, 1.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.0%, -0.8%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.9% | [-1.0%, -0.8%] | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b397bc0727ad27340466166455c6edd327a589c4&end=2cd2070af7643ad88d280a4933bc4fb60451e521&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.8%, -0.8%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-1.8%, -1.8%] | 1     |
| All ❌✅ (primary)                 | -1.8% | [-2.8%, -0.8%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b397bc0727ad27340466166455c6edd327a589c4&end=2cd2070af7643ad88d280a4933bc4fb60451e521&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.7% | [-1.9%, -1.6%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.7% | [-1.9%, -1.6%] | 4     |
</details>

nnethercote commented:
There is a single trivial regression, which may or may not be real.

@rustbot label: +perf-regression-triaged
